File: jacsgo/method_parity_test.go
package jacs

// Method enumeration parity test for the Go binding.
//
// Validates that all methods listed in
// binding-core/tests/fixtures/method_parity.json are exposed on the
// Go JacsSimpleAgent struct, with documented exclusions and Go-style
// name mappings.
//
// This is a *structural* test (method names), not a *behavioral* test.
// It complements, not duplicates, simple_agent_parity_test.go.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"sort"
	"testing"
)

const methodFixtureRelPath = "../binding-core/tests/fixtures/method_parity.json"

type methodParityFixture struct {
	AllMethodsFlat []string `json:"all_methods_flat"`
}

func loadMethodParityFixture(t *testing.T) methodParityFixture {
	t.Helper()

	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	fixturePath := filepath.Join(filepath.Dir(thisFile), methodFixtureRelPath)

	data, err := os.ReadFile(fixturePath)
	if err != nil {
		t.Fatalf("failed to read method_parity.json at %s: %v", fixturePath, err)
	}

	var f methodParityFixture
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("failed to parse method_parity.json: %v", err)
	}
	return f
}

// Methods that are intentionally not exposed in Go (with reasons).
var excludedFromGo = map[string]string{
	// inner_ref returns a raw Rust reference; not meaningful across CGo FFI
	"inner_ref": "raw Rust reference, not FFI-safe",
	// from_agent wraps a Rust SimpleAgent; not callable from Go
	"from_agent": "Rust-internal constructor",
	// load_with_info is an internal Rust helper; Go uses LoadSimpleAgent()
	"load_with_info": "internal helper, Go uses LoadSimpleAgent",
	// Conversion methods are not exposed via the CGo FFI layer.
	// They would require additional C wrapper functions.
	"to_yaml":     "not exposed via CGo FFI",
	"from_yaml":   "not exposed via CGo FFI",
	"to_html":     "not exposed via CGo FFI",
	"from_html":   "not exposed via CGo FFI",
	"rotate_keys": "not exposed via CGo FFI",
}

// Rust snake_case -> Go PascalCase method name mapping.
// Constructors are package-level functions, not methods on *JacsSimpleAgent.
var goNameMap = map[string]string{
	"create":              "NewSimpleAgent",          // constructor (package-level)
	"load":                "LoadSimpleAgent",         // constructor (package-level)
	"ephemeral":           "EphemeralSimpleAgent",    // constructor (package-level)
	"create_with_params":  "CreateSimpleAgentWithParams", // constructor (package-level)
	"get_agent_id":        "GetAgentID",
	"key_id":              "KeyID",
	"is_strict":           "IsStrict",
	"config_path":         "ConfigPath",
	"export_agent":        "ExportAgent",
	"get_public_key_pem":  "GetPublicKeyPEM",
	"get_public_key_base64": "GetPublicKeyBase64",
	"diagnostics":         "Diagnostics",
	"verify_self":         "VerifySelf",
	"verify_json":         "Verify",
	"verify_with_key_json": "VerifyWithKey",
	"verify_by_id_json":   "VerifyByID",
	"sign_message_json":   "SignMessage",
	"sign_raw_bytes_base64": "SignRawBytes",
	"sign_file_json":      "SignFile",
}

// Constructors are package-level functions, not methods on *JacsSimpleAgent.
var goConstructors = map[string]bool{
	"NewSimpleAgent":              true,
	"LoadSimpleAgent":             true,
	"EphemeralSimpleAgent":        true,
	"CreateSimpleAgentWithParams": true,
}

// goConstructorFuncs references actual constructor functions so the compiler
// catches removals. If any constructor is renamed or deleted, this file fails
// to compile -- no runtime test needed.
var goConstructorFuncs = map[string]interface{}{
	"NewSimpleAgent":              NewSimpleAgent,
	"LoadSimpleAgent":             LoadSimpleAgent,
	"EphemeralSimpleAgent":        EphemeralSimpleAgent,
	"CreateSimpleAgentWithParams": CreateSimpleAgentWithParams,
}

// TestMethodParityAgainstFixture validates that all expected methods exist
// on Go's JacsSimpleAgent.
func TestMethodParityAgainstFixture(t *testing.T) {
	fixture := loadMethodParityFixture(t)

	// Get all methods on *JacsSimpleAgent via reflection
	agentType := reflect.TypeOf(&JacsSimpleAgent{})
	instanceMethods := make(map[string]bool)
	for i := 0; i < agentType.NumMethod(); i++ {
		instanceMethods[agentType.Method(i).Name] = true
	}

	missing := []string{}
	for _, rustName := range fixture.AllMethodsFlat {
		if _, excluded := excludedFromGo[rustName]; excluded {
			continue
		}

		goName, mapped := goNameMap[rustName]
		if !mapped {
			missing = append(missing, rustName+" (no goNameMap entry)")
			continue
		}

		if goConstructors[goName] {
			// Constructors are package-level functions, verified at
			// compile time via goConstructorFuncs (references the actual
			// functions). If a constructor is removed, this file fails
			// to compile.
			continue
		}

		if !instanceMethods[goName] {
			missing = append(missing, rustName+" -> "+goName)
		}
	}

	if len(missing) > 0 {
		sort.Strings(missing)
		t.Errorf("Go JacsSimpleAgent is missing %d methods from method_parity.json:\n%s\n\n"+
			"If a method was intentionally excluded, add it to excludedFromGo.\n"+
			"If it has a different Go name, add it to goNameMap.",
			len(missing), formatLines(missing))
	}
}

// TestMethodParityExclusionsAreValid verifies every excluded method exists in the fixture.
func TestMethodParityExclusionsAreValid(t *testing.T) {
	fixture := loadMethodParityFixture(t)
	allMethods := make(map[string]bool)
	for _, m := range fixture.AllMethodsFlat {
		allMethods[m] = true
	}

	invalid := []string{}
	for excluded := range excludedFromGo {
		if !allMethods[excluded] {
			invalid = append(invalid, excluded)
		}
	}

	if len(invalid) > 0 {
		t.Errorf("excludedFromGo contains methods not in the fixture: %v. Remove stale exclusions.", invalid)
	}
}

// TestMethodParityNameMapCoversAll verifies the name map covers all non-excluded methods.
func TestMethodParityNameMapCoversAll(t *testing.T) {
	fixture := loadMethodParityFixture(t)

	unmapped := []string{}
	for _, rustName := range fixture.AllMethodsFlat {
		if _, excluded := excludedFromGo[rustName]; excluded {
			continue
		}
		if _, mapped := goNameMap[rustName]; !mapped {
			unmapped = append(unmapped, rustName)
		}
	}

	if len(unmapped) > 0 {
		t.Errorf("Methods without goNameMap entry: %v. Add a mapping.", unmapped)
	}
}

// TestMethodParityExclusionsAreStillNeeded checks if excluded methods
// have since been exposed on *JacsSimpleAgent. If a method was excluded
// because it wasn't in the CGo FFI layer but has since been added, this
// test fails to prompt removal of the exclusion.
func TestMethodParityExclusionsAreStillNeeded(t *testing.T) {
	agentType := reflect.TypeOf(&JacsSimpleAgent{})

	// Only check conversion-method exclusions (the ones likely to change).
	// internal-only exclusions (inner_ref, from_agent, load_with_info) will
	// never appear as Go methods.
	conversionExclusions := map[string]string{
		"to_yaml":  "ToYaml",
		"from_yaml": "FromYaml",
		"to_html":  "ToHtml",
		"from_html": "FromHtml",
	}

	newlyAvailable := []string{}
	for rustName, goName := range conversionExclusions {
		_, found := agentType.MethodByName(goName)
		if found {
			newlyAvailable = append(newlyAvailable, rustName+" -> "+goName)
		}
	}

	if len(newlyAvailable) > 0 {
		sort.Strings(newlyAvailable)
		t.Errorf("Excluded methods are now available on *JacsSimpleAgent. "+
			"Remove them from excludedFromGo and verify they work:\n%s",
			formatLines(newlyAvailable))
	}
}

func formatLines(items []string) string {
	result := ""
	for _, item := range items {
		result += "  - " + item + "\n"
	}
	return result
}

File: jacsnpm/test/method-parity.test.js
/**
 * Method enumeration parity test for the Node.js binding.
 *
 * Validates that all methods listed in
 * binding-core/tests/fixtures/method_parity.json are exposed on the
 * JacsSimpleAgent class, with documented exclusions and camelCase name mappings.
 *
 * This is a *structural* test (method names), not a *behavioral* test.
 * It complements, not duplicates, test_parity.js.
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { expect } = require('chai');

const FIXTURE_PATH = path.resolve(
  __dirname,
  '../../binding-core/tests/fixtures/method_parity.json'
);

// Methods that are intentionally Rust-only and not exposed in Node.
const EXCLUDED_FROM_NODE = new Set([
  // inner_ref returns a raw Rust reference; not meaningful across FFI
  'inner_ref',
  // from_agent wraps a Rust SimpleAgent; not callable from JS
  'from_agent',
  // load_with_info is an internal Rust helper; Node uses load() directly
  'load_with_info',
  // Gated on the `a2a` cargo feature, which the default Node build does not
  // enable (jacsnpm default features are attestation + agreements).
  'export_a2a_agent_card_json',
]);

// Rust snake_case method name -> Node camelCase method name mapping.
const NODE_NAME_MAP = {
  'create': 'create',               // static
  'load': 'load',                   // static
  'ephemeral': 'ephemeral',         // static
  'create_with_params': 'createWithParams', // static
  'get_agent_id': 'getAgentId',
  'key_id': 'keyId',
  'is_strict': 'isStrict',
  'config_path': 'configPath',
  'export_agent': 'exportAgent',
  'get_public_key_pem': 'getPublicKeyPem',
  'get_public_key_base64': 'getPublicKeyBase64',
  'diagnostics': 'diagnostics',
  'verify_self': 'verifySelf',
  'verify_json': 'verify',
  'verify_with_key_json': 'verifyWithKey',
  'verify_by_id_json': 'verifyById',
  'sign_message_json': 'signMessage',
  'sign_raw_bytes_base64': 'signRawBytes',
  'sign_file_json': 'signFile',
  'build_auth_header': 'buildAuthHeader',
  'build_request_auth_header': 'buildRequestAuthHeader',
  'canonicalize_json': 'canonicalizeJson',
  'sign_response': 'signResponse',
  'encode_verify_payload': 'encodeVerifyPayload',
  'decode_verify_payload': 'decodeVerifyPayload',
  'extract_document_id': 'extractDocumentId',
  'prepare_signed_event_replay_json': 'prepareSignedEventReplay',
  'unwrap_signed_event': 'unwrapSignedEvent',
  'to_yaml': 'toYaml',
  'from_yaml': 'fromYaml',
  'to_html': 'toHtml',
  'from_html': 'fromHtml',
  'rotate_keys': 'rotateKeys',
  // ES256 compatibility key + exports (P2 Tasks 002 / 003 / 004 / 004b).
  'add_compat_key_json': 'addCompatKey',
  'issue_compat_binding_json': 'issueCompatBinding',
  'export_compatibility_jwks_json': 'exportCompatibilityJwks',
  'export_compatibility_key_binding_json': 'exportCompatibilityKeyBinding',
  'export_ap2_mandate_json': 'exportAp2Mandate',
  'export_w3c_did': 'exportW3cDid',
  'export_w3c_did_document_json': 'exportW3cDidDocument',
  'export_w3c_agent_description_json': 'exportW3cAgentDescription',
  'generate_w3c_well_known_json': 'generateW3cWellKnown',
  'sign_w3c_request_json': 'signW3cRequest',
  'verify_w3c_request_json': 'verifyW3cRequest',
  // Inline text + media (Task 11): NAPI methods on JacsSimpleAgent strip the
  // _json suffix. Short aliases (signText/verifyText) also exist on the class
  // but the parity name is the suffix-stripped name shown here.
  'sign_text_file_json': 'signTextFile',
  'verify_text_file_json': 'verifyTextFile',
  'sign_image_json': 'signImage',
  'verify_image_json': 'verifyImage',
  'extract_media_signature_json': 'extractMediaSignature',
  // Agreement v2 (feature-gated in Rust, exposed by the default Node build).
  'create_agreement_v2_json': 'createAgreementV2',
  'apply_agreement_v2_json': 'applyAgreementV2',
  'sign_agreement_v2_json': 'signAgreementV2',
  'verify_agreement_v2_json': 'verifyAgreementV2',
  'detect_agreement_v2_branch_conflict_json': 'detectAgreementV2BranchConflict',
  'merge_agreement_v2_transcript_branches_json': 'mergeAgreementV2TranscriptBranches',
  'resolve_agreement_v2_branch_conflict_json': 'resolveAgreementV2BranchConflict',
  'export_agreement_v2_as_vc_json': 'exportAgreementV2AsVc',
};

// Static methods (on the class itself, not on instances)
const STATIC_METHODS = new Set(['create', 'load', 'ephemeral', 'createWithParams']);

describe('Node.js method enumeration parity', function () {
  let fixture;
  let JacsSimpleAgent;
  let agent;

  before(function () {
    if (!fs.existsSync(FIXTURE_PATH)) {
      console.log('  Skipping method parity tests - fixture not found');
      this.skip();
      return;
    }
    fixture = JSON.parse(fs.readFileSync(FIXTURE_PATH, 'utf8'));

    try {
      const bindings = require('../index.js');
      JacsSimpleAgent = bindings.JacsSimpleAgent;
      if (!JacsSimpleAgent) {
        this.skip();
        return;
      }
      agent = JacsSimpleAgent.ephemeral('ed25519');
    } catch (e) {
      console.log('  Skipping method parity tests - native binding not available');
      this.skip();
    }
  });

  function parityMethods() {
    const methods = [...fixture.all_methods_flat];
    for (const gated of Object.values(fixture.feature_gated_methods || {})) {
      methods.push(...gated);
    }
    return methods;
  }

  it('all non-excluded methods from fixture exist on JacsSimpleAgent', function () {
    const allMethods = parityMethods();
    const missing = [];

    for (const rustName of allMethods) {
      if (EXCLUDED_FROM_NODE.has(rustName)) continue;

      const nodeName = NODE_NAME_MAP[rustName];
      if (!nodeName) {
        missing.push(`${rustName} (no NODE_NAME_MAP entry)`);
        continue;
      }

      if (STATIC_METHODS.has(nodeName)) {
        // Check on the class itself
        if (typeof JacsSimpleAgent[nodeName] !== 'function') {
          missing.push(`${rustName} -> static ${nodeName}`);
        }
      } else {
        // Check on the instance
        if (typeof agent[nodeName] !== 'function') {
          missing.push(`${rustName} -> ${nodeName}`);
        }
      }
    }

    expect(missing, `Missing methods:\n${missing.join('\n')}`).to.be.empty;
  });

  it('public simple protocol helpers perform a strict roundtrip', function () {
    const legacy = agent.buildAuthHeader();
    expect(legacy).to.match(/^JACS /);
    const body = '{"include_test":false}';
    const header = agent.buildRequestAuthHeader(
      'POST',
      'https://hai.ai/api/v1/agents/hello',
      body,
      'hai.ai'
    );
    expect(header).to.match(/^JACS v2\./);

    const envelope = agent.signResponse('{"type":"connected"}');
    const parsed = JSON.parse(envelope);
    const signerId = parsed.jacsSignature.agentID;
    const verified = JSON.parse(agent.unwrapSignedEvent(
      envelope,
      JSON.stringify({ [signerId]: agent.getPublicKeyPem() })
    ));
    expect(verified.verified).to.equal(true);
    expect(verified.data.type).to.equal('connected');
  });

  it('request auth hashes exact Buffer and Uint8Array body bytes', function () {
    const backing = new Uint8Array([0xaa, 0x10, 0x00, 0xff, 0x20, 0xbb]);
    const bodies = [
      Buffer.from([0x00, 0xff, 0x62, 0x00, 0x79]),
      new Uint8Array([0x80, 0x00, 0xfe, 0x7f]),
      backing.subarray(1, 5),
    ];
    for (const body of bodies) {
      const header = agent.buildRequestAuthHeader(
        'POST',
        'https://hai.ai/api/v1/jobs',
        body,
        'hai.ai',
      );
      const claimsSegment = header.slice('JACS v2.'.length).split('.', 1)[0];
      const claims = JSON.parse(Buffer.from(claimsSegment, 'base64url').toString('utf8'));
      const expected = crypto.createHash('sha256').update(body).digest('base64');
      expect(claims.contentDigest).to.equal(`sha-256=:${expected}:`);
    }
  });

  it('exclusions are all valid fixture methods', function () {
    const allMethods = new Set(parityMethods());
    const invalid = [];
    for (const excluded of EXCLUDED_FROM_NODE) {
      if (!allMethods.has(excluded)) {
        invalid.push(excluded);
      }
    }
    expect(invalid, `EXCLUDED_FROM_NODE contains methods not in fixture: ${invalid}`).to.be.empty;
  });

  it('NODE_NAME_MAP covers all non-excluded methods', function () {
    const allMethods = parityMethods();
    const unmapped = [];
    for (const rustName of allMethods) {
      if (EXCLUDED_FROM_NODE.has(rustName)) continue;
      if (!(rustName in NODE_NAME_MAP)) {
        unmapped.push(rustName);
      }
    }
    expect(unmapped, `Methods without NODE_NAME_MAP entry: ${unmapped}`).to.be.empty;
  });

  it('NODE_NAME_MAP has no stale entries', function () {
    const allMethods = new Set(parityMethods());
    const stale = [];
    for (const rustName of Object.keys(NODE_NAME_MAP)) {
      if (!allMethods.has(rustName)) {
        stale.push(rustName);
      }
    }
    expect(stale, `Stale NODE_NAME_MAP entries: ${stale}`).to.be.empty;
  });

  it('method count matches fixture minus exclusions', function () {
    const expected = parityMethods().length - EXCLUDED_FROM_NODE.size;
    const nodeNameCount = Object.keys(NODE_NAME_MAP).length;
    expect(nodeNameCount).to.equal(expected,
      `NODE_NAME_MAP has ${nodeNameCount} entries but expected ${expected} (${EXCLUDED_FROM_NODE.size} excluded)`
    );
  });
});
